### PR TITLE
Implement #8057 103 Early Hint

### LINKED ISSUE #8057: Support Http Response 103 (Early Hints)
**Target Jetty version(s)**
10+

**Enhancement Description**

What would it take to have Jetty support response code 103 (Early Hints)?

* https://datatracker.ietf.org/doc/html/rfc8297
* https://www.fastly.com/blog/beyond-server-push-experimenting-with-the-103-early-hints-status-code
* https://developers.cloudflare.com/cache/about/early-hints/

### PR DESCRIPTION
**Pull request by gregw**
Implementing #8057 Early hints

### CODE REVIEW

**Comment by gregw** [comment]
@lorban can you look at why h3 is failing.

**Comment by lorban** [comment]
@gregw I see no H3 test failure in the PR's build, only a checkstyle error. What am I overlooking?

**Comment by gregw** [comment]
@lorban have you run org.eclipse.jetty.http.client.IntermediateResponseTest#test103EarlyHint ?

**Comment by lorban** [comment]
@gregw I've added the missing support in H3 for 103 status code and pushed the changes to this branch.

While your test now passes and I'm reasonably confident my changes are correct, it would still be wise to have @sbordet review this bit eventually.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java [hunk 1]
```diff
@@ -388,12 +388,18 @@ public Result generateResponse(MetaData.Response info, boolean head, ByteBuffer
                     if (status >= 100 && status < 200)
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java]: There is `HttpStatus.isInformational()` that we should use here.

File: jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public void sendHeaders(MetaData.Request request, MetaData.Response response, By
         boolean isHeadRequest = HttpMethod.HEAD.is(request.getMethod());
         boolean hasContent = BufferUtil.hasContent(content) && !isHeadRequest;
         int status = response.getStatus();
-        boolean interimResponse = status == HttpStatus.CONTINUE_100 || status == HttpStatus.PROCESSING_102;
+        boolean interimResponse = status >= HttpStatus.CONTINUE_100 && status < HttpStatus.OK_200 && status != HttpStatus.SWITCHING_PROTOCOLS_101;
```
Comment (sbordet) [inline jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java]: Replace with `HttpStatus.isInformational() && !101`.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    public void init(Transport transport) throws IOException
+    {
+        // Skip FCGI for now, not much interested in its server-side behavior.
+        Assumptions.assumeTrue(transport != FCGI);
+        setScenario(new TransportScenario(transport));
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(TransportProvider.class)
+    public void test100Continue(Transport transport) throws Exception
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: There are a lot of 100-Continue tests elsewhere, so I don't think we need to duplicate this here.
I'd remove this test.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                        org.eclipse.jetty.client.api.Request request = response.getRequest();
+                        HttpConversation conversation = ((HttpRequest)request).getConversation();
+                        // Reset the conversation listeners, since we are going to receive another response code
+                        conversation.updateResponseListeners(null);
+
+                        HttpExchange exchange = conversation.getExchanges().peekLast();
+                        if (exchange != null && response.getStatus() == HttpStatus.EARLY_HINT_103)
+                        {
+                            // All good, continue.
+                            System.err.println("onSuccess\n" + response.getHeaders());
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: Remove `System.err`

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                            hints.add(response.getHeaders().get("Hint"));
+                            exchange.resetResponse();
+                            exchange.proceed(null);
+                        }
+                        else
+                        {
+                            throw new IllegalStateException("should not have accepted");
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: This exception will only be logged, better call `response.abort(new ISE(...))`, so the test will fail.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                        HttpExchange exchange = conversation.getExchanges().peekLast();
+                        if (exchange != null && response.getStatus() == HttpStatus.PROCESSING_102)
+                        {
+                            // All good, continue.
+                            exchange.resetResponse();
+                            exchange.proceed(null);
+                        }
+                        else
+                        {
+                            throw new IllegalStateException("should not have accepted");
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: Use `response.abort(new ISE(...))`.

File: jetty-http3/http3-common/src/main/java/org/eclipse/jetty/http3/internal/HTTP3StreamConnection.java [hunk 7]
```diff
@@ -486,7 +486,7 @@ public void onHeaders(long streamId, HeadersFrame frame)
             else if (metaData.isResponse())
             {
                 MetaData.Response response = (MetaData.Response)metaData;
-                if (response.getStatus() != HttpStatus.CONTINUE_100)
+                if (response.getStatus() != HttpStatus.CONTINUE_100 && response.getStatus() != HttpStatus.EARLY_HINT_103)
```
Comment (sbordet) [inline jetty-http3/http3-common/src/main/java/org/eclipse/jetty/http3/internal/HTTP3StreamConnection.java]: This should be `HttpStatus.isInformational()`.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java [hunk 8]
```diff
@@ -103,6 +104,7 @@
         CONTINUE(CONTINUE_100, "Continue"),
         SWITCHING_PROTOCOLS(SWITCHING_PROTOCOLS_101, "Switching Protocols"),
         PROCESSING(PROCESSING_102, "Processing"),
+        EARLY_HINT(EARLY_HINT_103, "Early Hint"),
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:107]: Let's add also `HttpHeader.Link` from RFC 8288.

File: jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java [hunk 9]
```diff
@@ -53,6 +53,8 @@ public void onResponse(HeadersFrame frame)
         boolean valid;
         if (response.getStatus() == HttpStatus.CONTINUE_100)
             valid = validateAndUpdate(EnumSet.of(FrameState.INITIAL), FrameState.CONTINUE);
+        else if (response.getStatus() == HttpStatus.EARLY_HINT_103)
+            valid = validateAndUpdate(EnumSet.of(FrameState.INITIAL, FrameState.HEADER, FrameState.CONTINUE), FrameState.CONTINUE);
```
Comment (sbordet) [inline jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java]: I would rename `FrameState.CONTINUE` to `INFORMATIONAL`.

State HEADER is the state used to validate data frames (i.e. we can only receive a DATA frame if we are in HEADER state).
So I think the "from" states can only be INITIAL and CONTINUE(INFORMATIONAL).

File: jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/internal/HttpTransportOverHTTP3.java [hunk 10]
```diff
@@ -71,7 +71,7 @@ private void sendHeaders(MetaData.Request request, MetaData.Response response, B
         boolean isHeadRequest = HttpMethod.HEAD.is(request.getMethod());
         boolean hasContent = BufferUtil.hasContent(content) && !isHeadRequest;
         int status = response.getStatus();
-        boolean interimResponse = status == HttpStatus.CONTINUE_100 || status == HttpStatus.PROCESSING_102;
+        boolean interimResponse = status >= HttpStatus.CONTINUE_100 && status < HttpStatus.OK_200 && status != HttpStatus.SWITCHING_PROTOCOLS_101;
```
Comment (sbordet) [inline jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/internal/HttpTransportOverHTTP3.java]: Replace with `HttpStatus.isInformational() && !101`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/Response.java [hunk 11]
```diff
@@ -514,6 +517,23 @@ public void sendProcessing() throws IOException
...
+    /**
+     * Sends a 102-Processing response.
+     * If the connection is an HTTP connection, the version is 1.1 and the
+     * request has a Expect header starting with 102, then a 102 response is
+     * sent. This indicates that the request still be processed and real response
+     * can still be sent.   This method is called by sendError if it is passed 102.
+     *
+     * @throws IOException if unable to send the 102 response
+     * @see javax.servlet.http.HttpServletResponse#sendError(int)
+     */
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:528]: Javadoc is wrong.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+import static org.eclipse.jetty.http.client.Transport.FCGI;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class IntermediateResponseTest extends AbstractTest<TransportScenario>
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: Rename to `InformationalResponseTest`.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+
+            @Override
+            public Response.Listener getResponseListener()
+            {
+                return new Response.Listener()
+                {
+                    @Override
+                    public void onSuccess(Response response)
+                    {
+                        org.eclipse.jetty.client.api.Request request = response.getRequest();
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: Use `var request = ...`.

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
```
Comment (sbordet) [inline tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: Use `var request = ...`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java (hunk 3 above)
Comment (gregw) [reply tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: I'm not sure the other tests made sending the content actually conditional on receiving the 100, as I certainly could not find the magic needed in the HttpClient elsewhere.    So I'd be more inclined to delete the other tests and keep this one.  I'll look for any other differences

Comment (gregw) [reply tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java]: OK now I find the good 100 test!

**Comment by gregw** [comment]
@lorban @sbordet I've updated for review feedback.
However, there is still something racy either in the tests or the impl.   UNIX_DOMAIN frequently fails for me if run with other tests (the second 103 is skipped).  I have once seen H2 fail by the 103 being reported as the final response instead of the 200, but I can't reproduce.     Something is not quiet right :(

**Comment by gregw** [comment]
Hmmm and I've just had a flake of the 102 test for UNIX_DOMAIN as well... also not reproducible :(

**Comment by gregw** [comment]
@sbordet what does the client do if it is still processing the 102/103 response and the final (or next) response turns up?    On the server, we don't wait for anything from the client before sending the next response, but if I put in a 10ms sleep between the responses then I don't get any flakes.      Either we are writing over the responses or the client is processing the next one before the last is finished?

**Comment by gregw** [comment]
@sbordet I think the race is in the client.  If I put the 10ms sleeps in the server I pass every time. However, if I put a 100ms sleep in the onSuccess handling for the 1xx responses, then I fail more often than not, even for HTTP.

File: jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java (hunk 9 above)
Comment (lorban) [reply jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java]: I agree that `FrameState.CONTINUE` should be renamed to `FrameState.INFORMATIONAL`.

I'm wondering if this code could be generalized or if we have no choice but to stick to a `if (status == 103)` special case.

File: jetty-server/src/main/java/org/eclipse/jetty/server/Response.java [hunk 15]
```diff
@@ -514,6 +517,22 @@ public void sendProcessing() throws IOException
         }
     }
 
+    /**
+     * Sends a 103 Early Hint response.
+     *
+     * Send a 103 response as per <a href="https://datatracker.ietf.org/doc/html/rfc8297">RFC8297</a>
+     * This method is called by sendError if it is passed 103.
+     *
+     * @throws IOException if unable to send the 102 response
```
Comment (lorban) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Response.java]:
```diff
- if unable to send the 102 response
+ if unable to send the 103 response
```

**Comment by lorban** [comment]
All tests in `IntermediateResponseTest` eventually fail when ran enough times, even if the H3 and UNIX_DOMAIN ones are more sensible.

File: jetty-http3/http3-common/src/main/java/org/eclipse/jetty/http3/internal/HTTP3StreamConnection.java (hunk 7 above)
Comment (gregw) [reply jetty-http3/http3-common/src/main/java/org/eclipse/jetty/http3/internal/HTTP3StreamConnection.java]: Is that OK for a 101 switching protocols?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java (hunk 8 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:107]: Added in javadoc

File: jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java (hunk 9 above)
Comment (gregw) [reply jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java]: There is no special case for 102???  Let's hangout and walk through this logic.

**Comment by gregw** [comment]
@sbordet @lorban The flaky issue is somewhere below the generation of the response in the server and the parsing of it in the client.   I added some println's and for a good run I see:
```
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=200,h=2,cl=-1}
startResponse 200 OK
```
So each generated response is parsed (startResponse) in the client.

But for a bad run I see:
```
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=200,h=2,cl=-1}
startResponse 200 OK
```
So all the responses are generated, but one of them is not received.     It feels like something might be clearing a buffer when it should not?     

@lorban can you focus on the H2/H3 flakes and I'll keep looking at h1

**Comment by gregw** [comment]
oops sorry forced pushed after a rebase

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java [hunk 16]
```diff
@@ -252,6 +252,7 @@ private boolean parse()
 
             if (complete)
             {
+                // TODO this can clean a pipelined response after a 1xx
                 if (LOG.isDebugEnabled())
                     LOG.debug("Discarding unexpected content after response: {}", networkBuffer);
                 networkBuffer.clear();
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: @sbordet @lorban this is the cause of the UNIX_DOMAIN flake.  A pipelined response can be discarded after a 1xx response.    I'm not sure of the fix.

**Comment by gregw** [comment]
@sbordet I found and maybe fixed the flake in the HttpClient (discarding content after a 1xx), but would appreciate very careful review.
@lorban I can't get any flaky fails in h2 or h3? How about you?

**Comment by lorban** [comment]
@gregw I can make `HttpClientStreamTest.testFileUpload` on H3 fail about once every couple dozen runs (I'm aware of that failure and I already started investigating it) but I cannot manage to get any of the tests in `InformationalResponseTest` to fail for H2 or H3 anymore.

That's strange...

**Comment by sbordet** [comment]
@gregw looking

**Review by lorban — APPROVED** [review]

**Review by sbordet — APPROVED** [review]

**Pull request merged**